File: app/desktop/sidecar/api/download.go
package api

import (
	"fmt"
	"io"
)

// GetFileMeta returns metadata for a file (including salt for decryption).
func (c *Client) GetFileMeta(fileID string) (*FileMetaResponse, error) {
	var resp FileMetaResponse
	err := c.doJSON("GET", fmt.Sprintf("/api/files/%s/meta", fileID), nil, &resp)
	return &resp, err
}

// GetFileChunk downloads a single encrypted chunk with metadata headers.
func (c *Client) GetFileChunk(fileID string, index int) (*ChunkDownload, error) {
	path := fmt.Sprintf("/api/files/%s/chunks/%d", fileID, index)
	resp, err := c.doRaw("GET", path, nil, nil)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode >= 400 {
		body, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("download chunk %d: HTTP %d: %s", index, resp.StatusCode, string(body))
	}

	data, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("read chunk %d: %w", index, err)
	}

	return &ChunkDownload{
		Data:       data,
		SHA256:     resp.Header.Get("X-Chunk-SHA256"),
		Compressed: resp.Header.Get("X-Chunk-Compressed") == "true",
	}, nil
}

File: app/core/src/api/download.rs
//! Download-side control-plane calls — port of `sidecar/api/download.go` plus
//! the byos-direct locators endpoint.

use super::client::{ApiError, Client};
use super::types::{ChunkDownload, FileLocatorsResponse, FileMetaResponse};

impl Client {
    /// GET /api/files/{id}/meta
    pub async fn get_file_meta(&self, file_id: &str) -> Result<FileMetaResponse, ApiError> {
        self.send_json(|http, base| http.get(format!("{base}/api/files/{file_id}/meta")))
            .await
    }

    /// GET /api/files/{id}/chunks/{idx} — relay download. Integrity metadata
    /// arrives in the X-Chunk-* headers.
    pub async fn get_chunk(&self, file_id: &str, idx: i64) -> Result<ChunkDownload, ApiError> {
        let resp = self
            .send(|http, base| http.get(format!("{base}/api/files/{file_id}/chunks/{idx}")))
            .await?;
        let status = resp.status();
        if !status.is_success() {
            return Err(ApiError::Status {
                status: status.as_u16(),
                body: resp.text().await.unwrap_or_default(),
            });
        }
        let sha256 = header(&resp, "X-Chunk-SHA256");
        let compressed = header(&resp, "X-Chunk-Compressed") == "true";
        let data = resp.bytes().await?.to_vec();
        Ok(ChunkDownload {
            data,
            sha256,
            compressed,
        })
    }

    /// GET /api/files/{id}/locators — owner-only per-chunk platform locations
    /// for byos-direct downloads.
    pub async fn get_file_locators(&self, file_id: &str) -> Result<FileLocatorsResponse, ApiError> {
        self.send_json(|http, base| http.get(format!("{base}/api/files/{file_id}/locators")))
            .await
    }
}

fn header(resp: &reqwest::Response, name: &str) -> String {
    resp.headers()
        .get(name)
        .and_then(|v| v.to_str().ok())
        .unwrap_or_default()
        .to_string()
}
